### PR TITLE
chore(readme): Document queries

### PR DESCRIPTION
**Pull request by k88hudson**
This change is [Reviewable](https://reviewable.io/reviews/mozilla/activity-streams/589)

### CODE REVIEW

File: queries.md [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+## General notes
+
+## Highlights
+
+### Description of query
+
+The highlights query is composed of two parts:
+
+( A) Select **1 item** from History where:
```
Comment (pdehaan) [inline queries.md]: Not sure if this should be `( A )` (with a space after "A" -- to match `( B )` below), or if we want to tempt fate and make it an ordered list.

File: queries.md [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+* the date the bookmark was created (`bookmarkDateCreated`) is in the last 3 days
+* the bookmark has been visited no more than 3 times (`visit_count`)
+
+Items are ordered (descending) by the date the bookmark was created (`bookmarkDateCreated`), meaning that newer item will come first.
+
+( B ) Then, select up to **20 items** from History, where:
+
+* the item was NOT visited (`last_visit_date`) in the last 30 minutes
+* the `rev_host` is unique (no two results returned should have the same `rev_host`
+* the `rev_host` is NOT one of "www.google.com", "www.google.ca", "calendar.google.com", "mail.google.com", "mail.yahoo.com", "search.yahoo.com", "localhost", "t.co". "."
```
Comment (pdehaan) [inline queries.md]: Not sure if the "." between "t.co" and "." should be a "." or ",".

File: queries.md [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+* `www.wired.com`, visited 45 minutes ago
+* `www.cnn.com/some-article` visited 2 hours ago
+
+### Examples of items that would *not* satisfy the requirements
+
+* A link with a bookmark created 5 minutes ago that was visited 4 times *(reason: it was visited more than 3 times)*
+* A link with a bookmark created last month that was visited once *(reason: it is too old)*
+* `www.wired.com` visited 5 minutes ago *(reason: it was visited too recently)*
+* `mail.google.com` visited 2 hours ago *(reason: mail.google.com is in the blocklist)*
+* `www.cnn.com/article-2`, if `wwww.cnn.com/article-1` were already included in the query *(reason: only one link per rev_host are allowed)*
```
Comment (pdehaan) [inline queries.md]: nit: Add code format to `rev_host` to match section above.

File: queries.md [hunk 4]
```diff
@@ -0,0 +1,51 @@
+# Query documentation
+
+## Highlights
+
+### Description of query
+
+The highlights query is composed of two parts:
+
+( A ) Select **1 item** from History where:
```
Comment (oyiptong) [inline queries.md]: I would say "attempt to select" for the first spot. If there is no such bookmark, the first spot will be replaced by items in the second part of the query

File: queries.md [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+* the date the bookmark was created (`bookmarkDateCreated`) is in the last 3 days
+* the bookmark has been visited no more than 3 times (`visit_count`)
+
+Items are ordered (descending) by the date the bookmark was created (`bookmarkDateCreated`), meaning that newer item will come first.
+
+( B ) Then, select up to **20 items** from History, where:
+
+* the item was NOT visited (`last_visit_date`) in the last 30 minutes
+* the `rev_host` is unique (no two results returned should have the same `rev_host`
+* the `rev_host` is NOT one of "www.google.com", "www.google.ca", "calendar.google.com", "mail.google.com", "mail.yahoo.com", "search.yahoo.com", "localhost", "t.co", "."
```
Comment (oyiptong) [inline queries.md]: the list of blocked hosts might grow. should we use the rev_host blacklist elsewhere, we should have a special section in the documentation for this list.

File: queries.md [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+
+* the item has a bookmark
+* the date the bookmark was created (`bookmarkDateCreated`) is in the last 3 days
+* the bookmark has been visited no more than 3 times (`visit_count`)
+
+Items are ordered (descending) by the date the bookmark was created (`bookmarkDateCreated`), meaning that newer item will come first.
+
+( B ) Then, select up to **20 items** from History, where:
+
+* the item was NOT visited (`last_visit_date`) in the last 30 minutes
```
Comment (oyiptong) [inline queries.md]: There is also a `visit_count` requirement of <= 3

File: queries.md [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+
+* A link with a bookmark created 10 seconds ago that was only visited once
+* A link with a bookmark created 2 days ago that was visited three times
+* `www.wired.com`, visited 45 minutes ago
+* `www.cnn.com/some-article` visited 2 hours ago
+
+### Examples of items that would *not* satisfy the requirements
+
+* A link with a bookmark created 5 minutes ago that was visited 4 times *(reason: it was visited more than 3 times)*
+* A link with a bookmark created last month that was visited once *(reason: it is too old)*
```
Comment (oyiptong) [inline queries.md]: I think that link would satisfy the requirements of the B) part of the highlights query.

* it was last visited more than 3 days ago
* the visit count is <= 3

File: queries.md [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+* `www.wired.com` visited 5 minutes ago *(reason: it was visited too recently)*
```
Comment (oyiptong) [inline queries.md]: it would not satisfy the requirements if it didn't have a bookmark, yes

File: queries.md (hunk 6 above)
Comment (oyiptong) [inline queries.md]: maybe it would be clearer to say that the last visit has to be more than 30 mins ago

File: queries.md [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+* `mail.google.com` visited 2 hours ago *(reason: mail.google.com is in the blocklist)*
+* `www.cnn.com/article-2`, if `wwww.cnn.com/article-1` were already included in the query *(reason: only one link per `rev_host` are allowed)*
+
+### Additional notes
+
+Some requirements/adjustments are also made to the initial query after enhanced data is received from Embedly and the links are placed in the UI:
+
+* Links that are already in the Top Sites section are removed from Highlights
+* Links that have **images and good titles/descriptions** are prioritized are sorted before those that do not
+* Only links that **already have cached data available** are returned to the UI, meaning that not all links returned from the query will reach the UI.
```
Comment (oyiptong) [inline queries.md]: This line is incorrect as of 1f9f5a0a134a3eac770a9a00297fb72be53a85e9

The links will be returned even if there are no enhanced previews available

File: queries.md [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+
+### Possible empty conditions
+
+* A new profile or empty history with no bookmarks, in the first 30 minutes of usage.
+* The embedly cache is empty/all items returned do not yet have cached data
```
Comment (oyiptong) [inline queries.md]: This line is incorrect as of 1f9f5a0a134a3eac770a9a00297fb72be53a85e9

The line also pertains to the UI while the comment at line 50 pertains to the query.
If this documentation page is about the query itself, we should specify the conditions as the title states this is for the query

File: queries.md (hunk 4 above)
Comment (k88hudson) [inline queries.md]: 👍

**Comment by k88hudson** [comment]
Oops, I forgot about this PR -- added review updates

**Comment by k88hudson** [comment]
@SoftVision-PaulOiegas @SoftVision-VladBacia This is the documentation for the queries. Sorry, I totally forgot about it!

**Pull request closed without merging**